### PR TITLE
Support getting active API keys

### LINKED ISSUE #97995: GET API Key API filter out expired and/or invalidated API keys
### Description

The [Get API key information API](https://www.elastic.co/guide/en/elasticsearch/reference/master/security-api-get-api-key.html) can currently filter the returned keys by:
 * key id/name
 * keys owned by the calling user
 * keys owned by other users, by `realm_name` and `username`

Irrespective of the filtering used, both invalidated and expired (but not yet deleted) keys will be included in the response.
In the response, keys have the `invalidated` boolean property, as well as the `expiration` property that contains the UNIX timestamp past of which the key is no longer valid for authentication.

I propose we introduce a way for the API to support retrieving:
* only valid API keys (not return invalidated keys) - using a boolean request parameter
* keys not expired before a certain timestamp - using a timestamp parameter that defaults to "now" - also note that keys' experiation date is always checked against the local node's clock

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR adds a flag `active_only` to the [Get API key API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-api-key.html) to optionally retrieve active-only (i.e., neither expired nor invalidated) API keys. The flag defaults to false, i.e., by default, expired and invalidate API keys **are** included in the response (therefore we keep BWC). 

An example API call is:

```
GET /_security/api_key?active_only=true
```

The `active_only` flag can be used with realm and username based flags (e.g., `active_only=true&username=X`) as well as `name` and `id`. In the case it's used with `id` for an existing but non-active key the response code is 404. This is consistent with the `owner` flag behavior.

Closes https://github.com/elastic/elasticsearch/issues/97995

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 1]
```diff
@@ -1901,13 +1902,13 @@ public void getApiKeys(
...
             apiKeyIds,
-            false,
-            false,
+            activeOnly,
+            activeOnly,
             hit -> convertSearchHitToApiKeyInfo(hit, withLimitedBy),
             ActionListener.wrap(apiKeyInfos -> {
                 if (apiKeyInfos.isEmpty()) {
                     logger.debug(
-                        "No active api keys found for realms {}, user [{}], api key name [{}] and api key ids {}",
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:1910]: Removed `active` because it's misleading

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 2]
```diff
@@ -1901,13 +1902,13 @@ public void getApiKeys(
             username,
             apiKeyName,
             apiKeyIds,
-            false,
-            false,
+            activeOnly,
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:1905]: Only real change in API key service -- nice that our code was general enough to support this 🎉

**Comment by n1v0lg** [comment]
@elasticmachine update branch plz

**Review by ywangd — APPROVED** [review]
LGTM

File: x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/GetApiKeysRestIT.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.emptyArray;
+import static org.hamcrest.Matchers.equalTo;
+
+public class GetApiKeysRestIT extends SecurityOnTrialLicenseRestTestCase {
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/GetApiKeysRestIT.java:42]: Indeed. I'll add an issue to the backlog to refactor and consolidate API key test classes.

File: x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/GetApiKeysRestIT.java [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        createRole("manage_own_api_key_role", Set.of("manage_own_api_key"));
+        createUser(MANAGE_SECURITY_USER, END_USER_PASSWORD, List.of("manage_security_role"));
+        createRole("manage_security_role", Set.of("manage_security"));
+    }
+
+    public void testGetApiKeysWithActiveOnlyFlag() throws Exception {
+        final String apiKeyId0 = createApiKey(MANAGE_SECURITY_USER, "key-0");
+        final String apiKeyId1 = createApiKey(MANAGE_SECURITY_USER, "key-1");
+        // Set short enough expiration for the API key to be expired by the time we query for it
+        final String apiKeyId2 = createApiKey(MANAGE_SECURITY_USER, "key-2", TimeValue.timeValueNanos(1));
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/GetApiKeysRestIT.java:59]: It does feel like a bug, good catch! 

I haven't made up my mind yet about whether to fix it though. I can see some contrived scenario where customer is using `0`--or maybe even negative values if they're doing some date math--in e.g., some automated test pipeline and changing this breaks things for them. This feels very far-fetched though. 

On the other hand, the advantage of fixing this is also quite small. 

The only "real" scenario that comes to mind where someone might be using `0` or `-1` is if they're misinterpreting the API: `expiration=0` or `expiration=-1` could be interpreted as "never expire" -- if someone doesn't carefully read the docs, they might just assume that's what it means. But that's actually more of an argument for fixing the bug... 

TL;DR: I'm leaning (slightly) towards treating this as a bug, without BWC implications, and just fixing it, but I want to sit on this for a bit and see if I come up with a reason not to. What would be your preference?

For this PR, I will leave the `1ns` as is.

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/GetApiKeysRestIT.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/GetApiKeysRestIT.java:59]: My preference is also treating it as a bug since I cannot think of any useful thing it can do. Definitely not critical though because the ultimate behaviour is correct, i.e. the API key cannot be used. We may also want to get the team's consensus on this.
